src/tiled/stampbrush.cpp

```diff
@@ -386,6 +401,52 @@ void StampBrush::drawPreviewLayer(const QVector<QPoint> &list)
             mapDocument()->unifyTilesets(variation.map, mMissingTilesets);
 
             TileLayer *stamp = variation.tileLayer();
+            bool copyFlag = false;
+
+            //if staggered map, makes sure stamp stays the same
+            if ((mapDocument()->map()->orientation() == Map::Hexagonal
+                    || mapDocument()->map()->orientation() == Map::Staggered)
+                    && ((mapDocument()->map()->staggerAxis() == Map::StaggerY)?
+                         stamp->height() > 1 : stamp->width() > 1)) {
+                stamp = variation.tileLayer()->copy(variation.tileLayer()->bounds());
+                copyFlag = true;
+
+                if(mapDocument()->map()->staggerAxis() == Map::StaggerY) {
+                    int rowOfTop = p.y() - stamp->height() / 2;
+
+                    if ((variation.map->staggerIndex() == mapDocument()->map()->staggerIndex()
+                            && (rowOfTop&1) == 1)
+                               ||
+                           (variation.map->staggerIndex() != mapDocument()->map()->staggerIndex()
+                            && (rowOfTop&1) == 0)) {
+                        stamp->resize(QSize(stamp->width() + 1,stamp->height()),QPoint());
+
+                        for (int y = (variation.map->staggerIndex()+1)&1; y < stamp->height(); y += 2) {
+                            for(int x = stamp->width() - 2; x >= 0; --x) {
+                                stamp->setCell(x+1,y,stamp->cellAt(x,y));
+                                stamp->setCell(x,y,Cell());
```

Comment:
This line can still be moved out of the inner loop, since it only needs to be done for the left-most cell.